src/jobserver-posix.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <fcntl.h>
+#include <cassert>
+#include <cstring>
+#include <unistd.h>
+
+#include "jobserver.h"
+#include "util.h"
+
+Jobserver::Jobserver() : token_count_(0), fd_(-1)
+{}
+
+void Jobserver::Init() {
+  assert(fd_ < 0);
+
+  if (!ParseJobserverAuth("fifo")) {
+    return;
+  }
+
+  const char *jobserver = jobserver_name_.c_str();
+
+  fd_ = open(jobserver, O_NONBLOCK | O_CLOEXEC | O_RDWR);
+  if (fd_ < 0) {
+    Fatal("failed to open jobserver: %s: %s", jobserver, strerror(errno));
+  }
+
+  Info("Using jobserver: %s", jobserver);
+}
+
+Jobserver::~Jobserver() {
+  assert(token_count_ == 0);
+
+  if (fd_ >= 0) {
+    close(fd_);
+  }
+}
+
+bool Jobserver::Enabled() const {
+  return fd_ >= 0;
+}
+
+bool Jobserver::AcquireToken() {
+  char token;
+  int res = read(fd_, &token, 1);
```

Comment:
nit: for reliability loop around EINTR here (these can happen for arbitrary reasons in general).

Reply:
As far as I can tell: `EINTR` cannot happen on non-blocking file descriptors.

It is not actually documented anywhere, but it makes sense, and the code (`pipe_read()` in `fs/pipe.c` in Linux) confirms it.

I can add a `do {} while();` if you still want me to, but I'd rather avoid adding hard-to-test conditionals when there's no need.